### DISCUSSION TITLE
getStaticProps for non page components

### DISCUSSION THREAD

**Discussion by cernockyd** [vercel/next.js#13390] [Ideas]
**Update May 2022**: From @leerob.

Please read the [Layouts RFC](https://github.com/vercel/next.js/discussions/37136) – we'd love your feedback!

**Comment by jamesmosier** [comment]
`getStaticProps` must [always be called from a page](https://nextjs.org/docs/basic-features/data-fetching#only-allowed-in-a-page) located in `pages/*`

**Reply by cernockyd** [reply]
Hello, thank you for reply. Of course I know that :-) I was just thinking if there is some way to do same as getStaticProps for other components too. I don't think such feature is completely necessary, though.

**Reply by Gusterson** [reply]
I like the idea... I have a component that needs to call some special data at build time. I use that component across several pages and page types. Inserting the "call this particular components data" on every page that needs it violates DRY.  

Would be nice to have the fetch logic AND the component logic in one nice tiny bundle.

**Reply by Xcalytoe** [reply]
Thank you. I now know the reason why my getStaticProps was returning undefined for non page components.

**Reply by aleexcotton** [reply]
Same here

**Comment by rafaelalmeidatk** [comment]
This will be possible with suspense in the future, currently it is not available on the server yet. In the mean time you would have to use something like this: https://github.com/FormidableLabs/react-ssr-prepass. It would run through your React element tree and fire all the requests, then get their results that you can send as the page props (and possibly use with Context to send to the components). Keep in mind that this could be kinda slow depending on the side of the tree.

**Comment by ManAnRuck** [comment]
Such a feature would be awesome!

I have to use in a project a GraphQL API with limited depth 🤢
A feature to fetch static props on component level will solve a lot of issues for me 🤩

**Comment by nibtime** [comment]
Since Next 10, I have the impression, that this is pretty much the only thing left where I can express something with Gatsby that I can't express with Next and Next sort of just outran Gatsby in almost everything else.

So I would love to have an equivalent for `useStaticQuery` in Next 🤩. Then there would even be no GraphQL query detour (`() => Promise<T>` is actually a lot more versatile and better composable). And to be really brazen: have it also integrated with i18n routing somehow (locales are the only special case where I missed filtering capabilities of static queries when working with Gatsby). 

The Next community just has been adding features I really wanted lately, so I am quite confident this will sort of just pop up in an upcoming release if it is feasible in a good way 😄.

**Comment by leerob** [comment]
`getStaticProps` for individual components will be solved with [React Server Components](http://vercel.link/rsc)!

**Reply by kgatjens** [reply]
When can we expect to give it a try @leerob ? (React Server Components)

**Reply by patjagielski** [reply]
Any updates on this on when could expect this? @leerob

**Reply by vamshi9666** [reply]
Server components can be an alternative but not the equivalent. 
`getStaticProps` works with static export where as server components won't. 
Please correct me if i am wrong. 

I believe [gridsome](https://gridsome.org/docs/components/#add-graphql-to-components) does this already.

**Reply by handletastic** [reply]
are we expecting to have server components with getStaticProps with NextJS 12.1 ?

**Comment by matiadev** [comment]
Same here. I'm fetching translations for the navigation that's used across pages. I'm thinking that maybe I should write to a JSON file instead. I'm probably going to take the lazy route and send down the props. :laughing:

**Reply by JimLixin** [reply]
Any solution you go with eventually? I'm facing the same.

**Reply by patjagielski** [reply]
Funny enough I'm facing the same exact issue as well, I've been avoiding doing it this way but looks like I'll have to until the update comes

**Reply by mnai01** [reply]
> Funny enough I'm facing the same exact issue as well, I've been avoiding doing it this way but looks like I'll have to until the update comes

You can try to use next.js 12 with react server components but it's still in the experimental phase and so far I haven't had much success with it

**Reply by handletastic** [reply]
Yeah, I didn't have success with it. I would need to use getStaticProps on a layout to take advantage of ISR and revalidate at the layout level. I was hoping that server components would allow to use getStaticProps

Any suggestions / news regarding this?

https://nextjs.org/docs/basic-features/layouts

![image](https://user-images.githubusercontent.com/16025440/156063593-5b648536-e60e-4f31-90d1-5c9e68f7fc44.png)

**Reply by matiadev** [reply]
This has been a while ago but I had to send the props down.

**Comment by morewry** [comment]
Yeah, I need this, too. I have a component that renders on every page. There is some data that can be statically pre-rendered. It's really not feasible to repeat fetching the props necessary in `getStaticProps` on every single page.

I'd really love a hook whose purpose is to run at build time.

**Comment by ziogas** [comment]
I was struggling with the same problem repeatedly for specific components like header/footer. So while server components are WIP, here's the hacky but working solution I use now. Maybe it will help somebody 🙂 

**It fetches data during the build time only, and you can use it across any react component.** 

All with typescript and git-friendly way.

1. Declare a start script in package.json to work with nextjs locally:

`"start": "(npm run external-data:dev && next dev); npm run external-data:dev-cleanup"`

2. Declare the build script in package.json to fetch data upon build:

`"build": "npm run external-data && next build"`

3. Declare that external-data script that updates your data:

`"external-data": "npm run exec src/bin/external-data.ts"`

4. Declare a hacky workaround for development so that you can refresh the data and still keep working on your branch. Some git magic:

```
    "external-data:dev": "npm run exec:dev src/bin/external-data.ts && git update-index --skip-worktree -- src/data/*.json",
    "external-data:dev-cleanup": "git update-index --no-skip-worktree -- src/data/*.json && git co -- src/data/*.json",
```

5. Here's my `exec` and `exec:dev` scripts, needed as it calls ts-node. The Prod version doesn't perform type checks.
```
    "exec": "DOTENV_CONFIG_PATH=./.env.local ts-node -P tsconfig.standalone.json -r dotenv/config -r next/dist/server/node-polyfill-fetch.js -T",
    "exec:dev": "DOTENV_CONFIG_PATH=./.env.local ts-node -P tsconfig.standalone.json -r dotenv/config -r next/dist/server/node-polyfill-fetch.js"
```

The data files are stored in `src/data/*.json`. You can create the initial files, like `src/data/links.json`, and commit them to git. Add some dummy info with the correct data structure. Then, use those files as usual, like `import allLinks from 'data/links.json'`. 

6. Last but not least, here's the `external-data.ts` script barebones. It simply fetches data from the headless CMS and replaces needed json files. As it executes every time before you run `npm start`, or just before nextjs builds, you get the latest data from your cms.


```typescript
/* eslint-disable no-console */
import fs from 'fs';
import path from 'path';
import { getLinks } from 'lib/cms';
import { ILink } from 'interfaces';

const writeJsonFile = (filename: string, data: unknown): Promise<void> => {
  return fs.promises.writeFile(filename, JSON.stringify(data || []));
};

const getLinksData = async (basePath: string): Promise<void> => {
  let links: ILink[] = [];

  try {
    console.log('Fetching links...');
    links = await getLinks();
    console.log(`Fetched ${links?.length} links `);

    await writeJsonFile(path.join(basePath, 'src/data/links.json'), links || []);
  } catch (err) {
    console.error(err);
  }
};

export const main = async (): Promise<void> => {
  const basePath: string = process.cwd();
  await Promise.all([getLinksData(basePath)]);
};

if (module.children) {
  main();
}

const emptyDefault = {};

export default emptyDefault;

```

Is it hacky? Hell yes. Is it solving this specific nexjts limitation? Absolutely! 😅

**Reply by developerKumar** [reply]
How will you re-validate this data?

**Reply by Ansaen** [reply]
> 

Doesn´t that mean that u need to run yarn start for that to load everything ? It´s clever, but i would like to know if it´s possible to fetch data just like getStaticProps, but before build.

**Reply by ziogas** [reply]
Sorry for the delay; I didn't see the Github notification. The whole process is organized like this:

- The 1st command (npm/yarn start) isn't running build and is supposed to be running on the local development machine. It fetches the data, does git magic, and launches the next dev server. After the server is killed, it restores the original data files.
- The 2nd step (npm/yarn build) is `"build": "npm run external-data && next build"`, which ensures that the data is fetched just before the build in the production server. It isn't supposed to be used locally, only for the prod.

So if you don't need local fetch, use nextjs `next dev` for the local server. The re-validation happens only upon each build, so if your external data changes, you need some sort of webhooks support (for production). This flow still works well for me in various projects 🙂

**Reply by mukul-turing** [reply]
Hi everyone,

We can use [next-common-props](https://www.npmjs.com/package/next-common-props) library, to get data in getStaticProps at build time, and then load in any nested component just using a hook, without props drilling.

[Here](https://github.com/developerKumar/next-common-props-example) is the example using above library.

**Reply by jozef-drabik** [reply]
@ziogas Thanks for reply. I would like to ask one more questions it is somehow related to external-data. So when i try to "yarn build" it shows me this error: `FetchError: request to https://sportzoo-strapi-production.up.railway.app/api/footer?locale=en&populate=deep failed, reason: self signed certificate`

I know it´s coming from external data, but the solutions i tried didnt work for me.

![image](https://user-images.githubusercontent.com/104101610/174651940-0cdc187c-6f3e-4912-ae33-3fd0a32465c8.png)

**Reply by developerKumar** [reply]
> @ziogas Thanks for reply. I would like to ask one more questions it is somehow related to external-data. So when i try to "yarn build" it shows me this error: `FetchError: request to https://sportzoo-strapi-production.up.railway.app/api/footer?locale=en&populate=deep failed, reason: self signed certificate`
> 
> I know it´s coming from external data, but the solutions i tried didnt work for me.
> 
> ![image](https://user-images.githubusercontent.com/104101610/174651940-0cdc187c-6f3e-4912-ae33-3fd0a32465c8.png)

SSL certificate for above endpoint has been expired. That's why you are getting it.

**Reply by ziogas** [reply]
Yep, as developerKumar wrote, it's related to the SSL, not the example code

**Reply by jozef-drabik** [reply]
Thank you very much. By the way how can i make it work with internalization plugin from strapi ?

**Reply by ziogas** [reply]
Sorry, I have no clue, never used that
